### ISSUE TITLE
Support SSA/ASS styling

### ISSUE THREAD

**Issue by jakobkukla** [google/ExoPlayer#8435]
Add support for SSA/ASS styling. Currently there is only "minimal" support for ASS subtitles since styling was removed from the initial PR. (See https://github.com/google/ExoPlayer/pull/2281)

It would be great to have full support of the spec. Many subtitles use these styling features to work directly with the video (e.g. to substitute text in a video). SSA/ASS styling is also very common in certain genres of media like Cartoons, Anime or Karaoke.

**Comment by andrewlewis** [comment]
@icbaker Tentatively marking as an enhancement, but please dupe or close as needed. Thanks!

**Comment by icbaker** [comment]
Thanks for the request.

Unfortunately we don't currently have any capacity to improve ExoPlayer's SSA/ASS styling support. However we would welcome high quality PRs in this area.

I'm also happy to discuss various approaches with PR authors if a new feature doesn't fit easily into the existing code structure (either before the PR, or in the review, whichever is easier).

**Comment by 13steinj** [comment]
Could we determine a basic checklist of what isn't supported, to determine what features of the spec can be focused on in order of size and/or ease of implementation?

(I'm not familiar with what exactly even *is* supported currently).

**Comment by icbaker** [comment]
> Could we determine a basic checklist of what isn't supported, to determine what features of the spec can be focused on in order of size and/or ease of implementation?

That sounds like a good idea - the best way to do that is probably to compare the [`SsaDecoder`](https://github.com/google/ExoPlayer/blob/dev-v2/library/core/src/main/java/com/google/android/exoplayer2/text/ssa/SsaDecoder.java) to a copy of the spec. I've found the SSA/ASS spec to be fairly loose, and I'm not sure there's a canonical definition, but I've generally referred to this when working with these files in the past: http://moodub.free.fr/video/ass-specs.doc (linked from [Matroska's page on SSA](https://www.matroska.org/technical/subtitles.html#ssaass-subtitles)).

In case it's helpful, we would probably prioritise things roughly in this order:
* Features that look like they're implemented but don't work properly (basically bugs).
* Features that are similar to existing ones, and so can be implemented with small incremental changes.
* Features that don't fit into the existing code architecture - so implementing them requires significant refactoring.

I don't know if SSA has anything in the final category - but if it looks like it does then please come talk to us before spending ages refactoring `SsaDecoder`, as we might have some ideas about the best way to approach the problem.

**Comment by cosmo321** [comment]
I'm not a developer, so I can't exactly contribute a lot, but proper support of ass/ssa in Exoplayer is something I'd very much would like to have, as Plex for Android currently makes me sad...
I don't know if this makes sense in the context of Exoplayer, but it seems to me that the closest thing we have to a "standard" for ass/ssa is with libass, as a lot of notable players have implemented it and it is actively developed. I think at least being compliant with how libass renders subtitles would be a good idea.

**Comment by icbaker** [comment]
>I don't know if this makes sense in the context of Exoplayer, but it seems to me that the closest thing we have to a "standard" for ass/ssa is with libass, as a lot of notable players have implemented it and it is actively developed. I think at least being compliant with how libass renders subtitles would be a good idea.

Thanks for the pointer to libass, I took a quick look. I think you're possibly suggesting two alternative things:
* ExoPlayer could depend on libass and use it's logic for all the SSA rendering.
* People working on improving ExoPlayer's existing SSA rendering logic could use libass as a reference implementation to see how different pieces of content **should** be rendered, as an augmentation to the "spec".

In case anyone is considering trying the first of these, I'd like to pre-emptively warn against it. libass looks like it's written in C, and we're very reluctant to pull in native/JNI dependencies unless a pure Java implementation of the same behaviour is deemed ~impossible. Given SSA/ASS is a relatively simple plaintext-based format, it seems completely feasible to parse it fully using pure Java - so it's extremely unlikely we'd accept a PR that introduced a native dependency to achieve this. There's more discussion about (not) using native deps for parsing subtitles in #7746 and #7749.

The second suggestion (use it as a reference implementation) makes perfect sense - and afaict it's used by VLC, so it should be enough to just play any content with VLC and compare to ExoPlayer's behaviour.

**Comment by cosmo321** [comment]
I leave that up to actual developers to decide. I just thought I'd point out that libass exists and is probably the best reference we have, and that rendering subs differently from libass could potentially make people like me wonder why the subs behave differently in ExoPlayer. The second alternative is probably the better option for ExoPlayer. :)

**Comment by szaboa** [comment]
hey @icbaker , I'll check what the current `SsaDecoder` supports and what is missing in regards of styling, based on the specs you linked above, then I could probably start and implement those step by step (with my limited time).

@jakobkukla did you have anything specific in mind? I mean is there a specific styling option which you would like to see first?

**Comment by icbaker** [comment]
Thanks for taking a look! Breaking down the work into small PRs definitely sounds good.

**Comment by jakobkukla** [comment]
@szaboa To be honest I'm not really aware what is currently supported and what not. But the ability to display fonts and colors properly would be nice!

**Comment by szaboa** [comment]
@icbaker 
I've checked out and it seems that only the v4+ "alignment" is supported. Here is the list of all the style attributes of v4 and v4+ from the specs. Do you see specific styles that would be very cumbersome to implement at first sight? E.g. thinking of "fontname", and "encoding"..

Style name | v4 | v4+ | Note | Format line support | Override support | PR
-- | -- | -- | -- | -- | -- | -- 
Name | ✅ | ✅ | - | ✅ | - | -
Fontname | ✅ | ✅ | - | ❌ | ❌ | -
Fontsize | ✅ | ✅ | - | ✅ | ❌ | #8615 
PrimaryColour | ✅ | ✅ | v4 BBGGRR, v4+ AABBGGRR | ✅  | ❌ | #8490 
SecondaryColour | ✅ | ✅ |  v4 BBGGRR, v4+ AABBGGRR | ❌ | ❌ | -
OutlineColour | ✅ | ✅ | v4 BBGGRR, v4+ AABBGGRR <br> Called OutlineColour in v4+, and TertiaryColour in v4 | ❌ | ❌ | -
BackColour | ✅ | ✅ | v4 BBGGRR, v4+ AABBGGRR | ❌ | ❌ | -
Bold | ✅ | ✅ | - | ✅   | ❌ | #8654
Italic | ✅ | ✅ | - | ✅  | ❌ | #8654
Underline | ❌  | ✅  | - | ✅  | ❌ | #8851
Strikeout | ❌  | ✅  | - |  ✅  | ❌ | #8851
ScaleX | ❌  | ✅  | - | ❌ | ❌ | -
ScaleY | ❌  | ✅  | - | ❌ | ❌ | -
Spacing | ❌  | ✅  | - | ❌ | ❌ | -
Angle | ❌  | ✅  | - | ❌ | ❌ | -
BorderStyle | ✅ | ✅ | - | ❌ | ❌ | -
Outline | ✅ | ✅ | - | ❌ | ❌ | -
Shadow | ✅ | ✅ | - | ❌ | ❌ | -
BorderStyle | ✅ | ✅ | - | ❌ | ❌ | -
Alignment | ✅ | ❌ | values may be 1=Left, 2=Centered, 3=Right. Add 4 to the value for a "Toptitle". Add 8 to the value for a "Midtitle" | ❌ | ❌ | -
Alignment | ❌ | ✅ | layout of the numpad (1-3 sub, 4-6 mid, 7-9 top) | ✅ | ✅ | -
MarginL, MarginR, MarginV | ✅ | ✅ | - | ❌ | -  | #10169
AlphaLevel | ❌  | ✅  | - | ❌ | ❌ | -
Encoding | ✅ | ✅ | - | ❌ | ❌ | -


My plan is to start implement some easy ones first :)

**Comment by icbaker** [comment]
The spec I linked above (https://github.com/google/ExoPlayer/issues/8435#issuecomment-756651245) seems to think that Underline, Strikeout, ScaleX, ScaleY, Spacing and Angle **are** supported in v4+. In fact are possibly only supported there and not in v4. Maybe those rows in your table are accidentally flipped? Or maybe I'm misreading the spec...

The `SsaDecoder` currently ignores v4 styling completely. It's not clear if that's sensible - I don't know how prevalent each version is in the wild.

If you're going to add v4 support we probably need to do think about whether we can reuse the `SsaStyle` class or split it into two. It's up to you, but for now imo it would be easiest to focus on v4+ only.

These all look like they should be relatively easy (I've referenced the relevant Android styling span where it seems obvious):

* Fontname (`TypefaceSpan`)
* PrimaryColour (`ForegroundColorSpan`)
* BackColour (`BackgroundColorSpan`)
* Bold (`StyleSpan`)
* Italic (`StyleSpan`)
* Underline (`UnderlineSpan`)
* Strikeout (`StrikethroughSpan`)
* AlphaLevel (read together with PrimaryColour, though the spec I'm reading says "SSA does not use this yet" so maybe not worth bothering with)
* MarginL, R and V (in fact there's already a TODO for these :)). These will feed into Cue's position & line fields. This might be a bit messy - suggest you don't start here.

The spec I'm looking at doesn't really define these clearly enough to be sure:
* Fontsize (I suspect you can use `RelativeSizeSpan` or `AbsoluteSizeSpan` but I don't know which is relevant)
     * EDIT: Small correction - if we work out how the spec defines this, we can use `Cue.textSize` and `textSizeType` for encoding size information from the "Format" section of the file that applies to the whole cue. We only need to use spans for overrides that affect substrings of the cue text.

These look harder. They have basically no support in the existing `Cue` interchange format:
* SecondaryColour
* OutlineColour
* ScaleX and ScaleY
* Spacing
* Angle
* BorderStyle
* Outline
* Shadow

These are hard for other reasons:
* Encoding - I'm not sure how you're supposed to read this since it's in a text file - so you've already had to guess the encoding (right?). Currently we assume it's always UTF-8.

**Comment by szaboa** [comment]
Yep, you are right, the Underline, Strikeout, ScaleX, ScaleY, Spacing and Angle are supported in v4+ only, and not in v4. I will correct the table.

Thanks for the tips, will definitely start with the easy ones.

**Comment by szaboa** [comment]
Initial PR for "Primary Colour" is here #8490.

**Comment by moneytoo** [comment]
Is there any option/API to disable, customize or detect the styling? If an app sets custom `CaptionStyleCompat`, it now accepts all styling while the text color is now variable based on line from subtitle file. Also, how about when user uses platform `CaptioningManager` to customize subtitle look - again, text color is now variable. With static background/border colors and variable text color, the subtitle line may end up low contrast on unreadable (right...?). Does my thinking here make any sense? (I've never used SSA/ASS outside some testing.)

**Comment by icbaker** [comment]
@moneytoo Assuming you're using the ExoPlayer UI components, you can force the `SubtitleView` to ignore 'embedded' styles (i.e. those encoded in subtitle files) by calling [`subtitleView.setApplyEmbeddedStyles(false)`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/ui/SubtitleView.html#setApplyEmbeddedStyles(boolean)). You can get the `SubtitleView` from [`playerView.getSubtitleView()`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/ui/PlayerView.html#getSubtitleView()).

**Comment by szaboa** [comment]
I just want to note that I haven't forgotten about this, just need to find time to continue it :) 
Updated the table above with the PrimaryColour support.

**Comment by szaboa** [comment]
Initial PR for "FontSize" is here: #8615

**Comment by viggy96** [comment]
I don't understand how this is a 'low priority'. Proper SSA/ASS subtitle support is crucial to a video player.

**Comment by Weevild** [comment]
What's the latest update? And would this progress and (hopefully) final integration of ass work with the Nvidia Shield TV?

**Comment by icbaker** [comment]
> What's the latest update? And would this progress and (hopefully) final integration of ass work with the Nvidia Shield TV?

The table in the [comment above](https://github.com/google/ExoPlayer/issues/8435#issuecomment-762449001) has the latest status (it's been updated with PRs merged since then).

I don't see any reason this wouldn't work on Nvidia Shield.

**Comment by szaboa** [comment]
I plan to continue with the contributions on this, just haven't had too much free time lately. Is there any preferred priority regarding the styling features?

**Comment by cosmo321** [comment]
No preference on priority, really. Glad to hear it's not forgotten. :)

**Comment by Weevild** [comment]
> I don't see any reason this wouldn't work on Nvidia Shield.

It works but when using home media services such as Plex a transcode is always guaranteed, I'd assume this is because ASS isn't fully supported. Plex could probably do something about this at their end but that doesn't seem like an active inquiry.

**Comment by 13steinj** [comment]
@Weevild ASS isn't fully supported on Plex for android, ironically, because ExoPlayer doesn't support ASS yet, and Plex devs don't want to go through the effort themselves. It's actually quite silly because they could transcode the ASS subtitles to a different format rather than the entire video, but don't. Even when this issue is resolved on ExoPlayer's end it'll take some time for Plex to pull it in.

**Comment by Weevild** [comment]
> Plex devs don't want to go through the effort themselves.

Not a shocker there, but one could only wish. I switched to emby because they support virtually every format that Plex doesn't (subtitle-wise), but to have native support would just be a relief knowing my media is supported regardless.
> transcode the ASS subtitles to a different format rather than the entire video

I think emby even does this too, guess they've got their hands full on that free content.

**Comment by albino1** [comment]
@sixones what's needed from [szaboa's list](https://github.com/google/ExoPlayer/issues/8435#issuecomment-762449001) to get this working well enough to merge ASS support into Plex on Android devices?  If there's some obvious priority thing I think it would be beneficial for a lot of people if it gets tackled first.

**Comment by sixones** [comment]
@albino1 Probably not the right place for this, but Plex on Android does support ASS subtitles without the styling already (be sure to set your subtitle burn option to only image formats). We are working on full ASS support with positioning, styling and animations at the moment. If we transcode them into another subtitle format they would end up being the same as the basic level of support which we already offer - without the burn in option changed we burn them into the video to keep the full ASS styling with animations, styles and positions above the video.

**Comment by albino1** [comment]
Thanks for the reply, makes perfect sense.  Appreciate everything you do!

**Comment by an-ghinid** [comment]
@sixones is there an issue open on your github to track that? Very keen on avoiding ASS transcoding as it's the only thing keeping me from running PMS entirely on my NAS rather than on my Shield TV (and would keep me paying for Plex Pass for the hardware acceleration support that allows this)

**Comment by cosmo321** [comment]
> We are working on full ASS support with positioning, styling and animations at the moment. 

@sixones I would've loved to see some more information about that in one of the relevant threads on the Plex forums. Proper ASS support is so long overdue I can't even.

**Comment by Weevild** [comment]
@cosmo321 Indeed. It's the sole reason I switched over to Emby, because my content would always get transcoded due to the lack of support. I've been waiting on a progress-update.

**Comment by CasperNiels** [comment]
I hope this one hasn't been abandoned, to me the main benefit of ASS over basic SRT subtitles is the positioning, so I guess Margin L, R, V should be a priority.

**Comment by fixel99** [comment]
+1 to still really needing/wanting support for this

**Comment by szaboa** [comment]
No promises as have very limited time, but I'll try to look into the Margin L, R, V thingy on the weekend.

**Comment by lsl330** [comment]
> No promises as have very limited time, but I'll try to look into the Margin L, R, V thingy on the weekend.

thank you for continuing to improve the ass's subtitle, in addition, I found that the bilingual (two-line) ass subtitle support also has problems, many times can only show one language (one line)

**Comment by szaboa** [comment]
PR with Margin L, R, V support coming soon.

**Comment by szaboa** [comment]
@icbaker PR for `MarginL`, `MarginR`, `MarginV` support: https://github.com/google/ExoPlayer/pull/10169.

**Comment by szaboa** [comment]
@lsl330 
> thank you for continuing to improve the ass's subtitle, in addition, I found that the bilingual (two-line) ass subtitle support also has problems, many times can only show one language (one line)

I can't find any information about bilingual subtitles in the SSA/ASS specification, do you have more information on this by any chance? Also could you please provide a sample `.ass` sub to check it out meanwhile?

**Comment by sixones** [comment]
> I can't find any information about bilingual subtitles in the SSA/ASS specification, do you have more information on this by any chance? Also could you please provide a sample .ass sub to check it out meanwhile?

There's no bilingual support as such in ASS/SSA, but you can show multiple subtitle lines at the same time, usually at the top and bottom of the screen, some files can use this for different languages (usually for song lyrics).

**Comment by lsl330** [comment]
@szaboa 

> I can't find any information about bilingual subtitles in the SSA/ASS specification, do you have more information on this by any chance? Also could you please provide a sample `.ass` sub to check it out meanwhile?

here is the sample ass file，it can show Chinese and japanese subtitles，but in exoplayer，can only show the chinese subtitles。
[sample ass.zip](https://github.com/google/ExoPlayer/files/8471754/sample.ass.zip)

![PC](https://user-images.githubusercontent.com/54297603/162932091-99387c24-69c3-4815-934c-db88f5fcfa0e.jpg)
![Android](https://user-images.githubusercontent.com/54297603/162932128-0ef367be-6ef9-40e8-91bf-e1ca706f751c.jpg)

thank you！

**Comment by szaboa** [comment]
@lsl330 I have just tested the subtitle you provided, basically there are multiple lines with the same or overlapping times, as @sixones mentioned. This is supported in ExoPlayer from version 2.11.0 (2019-12-11). It has been solved here https://github.com/google/ExoPlayer/issues/6320. Which version are you using?

... however, in your case these lines have the same position, which is not handled so they are drawn on top of each other.
<img width="586" alt="Screenshot 2022-04-12 at 22 17 08" src="https://user-images.githubusercontent.com/3526708/163046393-1ea7808b-cc87-4cdb-bdba-fcef5dee70bf.png">

The SSA/ASS specs mentions two kind of collision handling, `normal` and `reverse`, I'll have a look on this.

**Comment by lsl330** [comment]
@szaboa 

OK,thank you very much!!!
I use plex and the exoplayer version is 2.17.1.

**Comment by david2tm** [comment]
Hi,
I'm not sure if it's a right place to write this -- but I want to try some coding related to SSA/ASS.
The list of stuff these subtitles can do is extensive. So I started looking at the code (mind it is the first time I'll try coding for Android), but I got env set-up already -- with example of subtitles which are not working as they should (looking at VLC rendering) and I want to make them work.
I wander, if it's possible to use WebView to render these Cue's?
I see that there's `WebViewSubtitleOutput`... so I guess it's already something.
Am I on the right path?

**Comment by david2tm** [comment]
@icbaker I looked into what SSA/ASS supports and what is implemented and what not. So what I think about it.
Currently there are two "subtitle" renderers. CanvasSubtitleOutput (the default) and WebViewSubtitleOutput (which I'm not sure how it's launched, I guess it's done by the app which uses ExoPlayer).
Anyway - to play with it I forced the WebViewSubtitleOutput, just because doing some styling is easier for me personally with CSS.
Many features ASS supports are not so easily done with CSS as well. I even played with doing stuff like BorderStyle, Outline, Shadow with SVGs, as CSS doesn't support BorderStyle-Outline the way ASS requires it.
And I'm not even talking about Effects -- which are basically animations.
I then went further down wanting to check how libass works and came across [their project](https://github.com/libass/JavascriptSubtitlesOctopus), which uses webassembly. Btw according to libass github page, Crunchyroll is using this project. I even entertained myself with using said webassembly with WebViewSubtitleOutput, but it's a no go.
As it stands - ExoPlayer looks at subtitles as a series of cues being rendered on a WebView and/or Canvas. Libass creates just another "video" stream based on description of text file. This makes sense. Check [this example page](https://libass.github.io/JavascriptSubtitlesOctopus/videojs.html). It showcases some capabilities of SSA/ASS which I'm not sure are at all possible with cues. 
So what are next steps?

1. One can hope that someone ports libass to Java -- which, probably will never happen :)

2. Quoting you from above. I now think that parsing is indeed easy, but rendering is next to impossible (with cues).
> libass looks like it's written in C, and we're very reluctant to pull in native/JNI dependencies unless a pure Java implementation of the same behaviour is deemed ~impossible. Given SSA/ASS is a relatively simple plaintext-based format, it seems completely feasible to parse it fully using pure Java

3. Keep it as it is -- as in, not ever adding the full list of features.

wdyt?

**Comment by buliaoyin** [comment]
Re-implement SSA/ASS render would be a pure way, but takes much more time to get full feature support, I think import libass with JNI would be the easiest way to make things done.

**Comment by jsravn** [comment]
Plex has now side stepped this by bundling libass in their android client, presumably using ffmpeg or something to mux on the client side before passing to exoplayer.

**Comment by mugiritsu** [comment]
> Re-implement SSA/ASS render would be a pure way, but takes much more time to get full feature support, I think import libass with JNI would be the easiest way to make things done.

That would be not only the easiest way, but also the only way to correctly render all those obscure and advanced feature-rich ASS files that typesetting enthusiasts can come up with. There are a lot .ass files with over 25MB of animated signs, and at this point not even those long-standing libass alternatives like vsfilter or its derivates can render them properly. Trying to recreate libass would be a neverending task and a huge magnet for complaints about any kind of advanced signs not rendering correctly. Not to mention, libass has years of performance optimizations.

Just adopting libass instead of making a new subtitle renderer could solve a lot of problems.

**Comment by Sparh4wk** [comment]
I would love to vouch for this one as well.. this is awesome player, but its rly bad for anime without full support for ssa subtitles :( 
Please make this happen.

**Comment by GlassedSilver** [comment]
> > Re-implement SSA/ASS render would be a pure way, but takes much more time to get full feature support, I think import libass with JNI would be the easiest way to make things done.
> 
> That would be not only the easiest way, but also the only way to correctly render all those obscure and advanced feature-rich ASS files that typesetting enthusiasts can come up with. There are a lot .ass files with over 25MB of animated signs, and at this point not even those long-standing libass alternatives like vsfilter or its derivates can render them properly. Trying to recreate libass would be a neverending task and a huge magnet for complaints about any kind of advanced signs not rendering correctly. Not to mention, libass has years of performance optimizations.
> 
> Just adopting libass instead of making a new subtitle renderer could solve a lot of problems.

I wholeheartedly agree.

If this issue hadn't been such a long-standing one and been actually looked into by Google itself rather than them watching from the sidelines and hoping for the community to fully do this job by the looks of it, I'd be a lot more optimistic about exoplayer re-implementing and keeping up with changes in-time or mostly in-time. But apparently relying on constantly staying in-sync with ass extensions is a futile effort, so just using libass seems like the most reliable, future-proof-ish effort there is from my limited understanding.

It's a shame, but the dynamic nature of ass is both a curse, but also a blessing because it offers features not found elsewhere. The cost may be annoying. But when libass is doing the work needed why don't we just rely on it like what I think any sensible player does with proper ass-support?

Why re-invent the wheel? It's not like there's much to gain from it right? Even if using libass might not be as performant as a fully-optimized re-implementation, what sorts of differences that can be felt are we talking about here? Subtitle rendering isn't something that needs to scale endlessly, especially not when it's per player, at humanly digestible play-speeds, etc...

To anyone who is investing any sort of work into this effort: you have my most honest gratitude. The day Android has player-independent ACE ass support will be a good day! Not only for us anime and FOSS app fans.

**Comment by FongMi** [comment]
Is there way to fix overlapping subtitles ?

**Comment by GlassedSilver** [comment]
> Is there way to fix overlapping subtitles ?

It's a known issue, any fix would be supplied by exoplayer extending ass support. I'm not aware of user settings that can fix this.

**Comment by Sparh4wk** [comment]
so its been two years already and we still dont have proper ssa support.. if using libass can solve this, can some dev do something about it? I think a lot of ppl would be rly happy. I know you are doing this for free, but still. thanks

**Comment by happyTonakai** [comment]
Is there any progress on it? Many other multimedia apps are based on ExoPlayer and thus have limited support on ass/ssa subtitles.

**Comment by FongMi** [comment]
> Is there any progress on it? Many other multimedia apps are based on ExoPlayer and thus have limited support on ass/ssa subtitles.

https://github.com/google/ExoPlayer/pull/10169

**Comment by happyTonakai** [comment]
> > Is there any progress on it? Many other multimedia apps are based on ExoPlayer and thus have limited support on ass/ssa subtitles.
> 
> #10169

I'm not sure if this can be called as a progress since nearly two years have passed and this pr is still not merged. In addition, there are still many styles that are not supported. Low priority label makes me desperate...

**Comment by e1ker** [comment]
I'd love to get this implemented as it would make this player the only one with DV and proper .ass support in Android besides Kodi which is way too heavy for anything other than an Nvidia Shield.

**Comment by zhuhang-jasper** [comment]
Looking forward to this!

**Comment by peerless2012** [comment]
L

> [@icbaker](https://github.com/icbaker) I looked into what SSA/ASS supports and what is implemented and what not. So what I think about it. Currently there are two "subtitle" renderers. CanvasSubtitleOutput (the default) and WebViewSubtitleOutput (which I'm not sure how it's launched, I guess it's done by the app which uses ExoPlayer). Anyway - to play with it I forced the WebViewSubtitleOutput, just because doing some styling is easier for me personally with CSS. Many features ASS supports are not so easily done with CSS as well. I even played with doing stuff like BorderStyle, Outline, Shadow with SVGs, as CSS doesn't support BorderStyle-Outline the way ASS requires it. And I'm not even talking about Effects -- which are basically animations. I then went further down wanting to check how libass works and came across [their project](https://github.com/libass/JavascriptSubtitlesOctopus), which uses webassembly. Btw according to libass github page, Crunchyroll is using this project. I even entertained myself with using said webassembly with WebViewSubtitleOutput, but it's a no go. As it stands - ExoPlayer looks at subtitles as a series of cues being rendered on a WebView and/or Canvas. Libass creates just another "video" stream based on description of text file. This makes sense. Check [this example page](https://libass.github.io/JavascriptSubtitlesOctopus/videojs.html). It showcases some capabilities of SSA/ASS which I'm not sure are at all possible with cues. So what are next steps?
> 
> 1. One can hope that someone ports libass to Java -- which, probably will never happen :)
> 2. Quoting you from above. I now think that parsing is indeed easy, but rendering is next to impossible (with cues).
> 
> > libass looks like it's written in C, and we're very reluctant to pull in native/JNI dependencies unless a pure Java implementation of the same behaviour is deemed ~impossible. Given SSA/ASS is a relatively simple plaintext-based format, it seems completely feasible to parse it fully using pure Java
> 
> 3. Keep it as it is -- as in, not ever adding the full list of features.
> 
> wdyt?

Good news, the libass for exoplayer is on the way, please check our progress in [libass-android](https://github.com/peerless2012/libass-android)

**Comment by peerless2012** [comment]
> Is there any progress on it? Many other multimedia apps are based on ExoPlayer and thus have limited support on ass/ssa subtitles.

You can find the libass for exoplayer at [libass-android](https://github.com/peerless2012/libass-android)

**Comment by icbaker** [comment]
> As it stands - ExoPlayer looks at subtitles as a series of cues being rendered on a WebView and/or Canvas. Libass creates just another "video" stream based on description of text file.

This is why I suggested it might fit better to use `libass` to drive a custom effect (https://github.com/androidx/media/issues/2053#issuecomment-2601704056) - my intuition is that plumbing the subtitle data into a custom effect will be easier than trying to pass video-like data through ExoPlayer's `Cue` machinery.

**Comment by peerless2012** [comment]
> > As it stands - ExoPlayer looks at subtitles as a series of cues being rendered on a WebView and/or Canvas. Libass creates just another "video" stream based on description of text file.
> 
> This is why I suggested it might fit better to use `libass` to drive a custom effect ([androidx/media#2053 (comment)](https://github.com/androidx/media/issues/2053#issuecomment-2601704056)) - my intuition is that plumbing the subtitle data into a custom effect will be easier than trying to pass video-like data through ExoPlayer's `Cue` machinery.

Libass produce bitmap as output, witch is compat for Cue with bitmap. I already finish this with ARGB_888 bitmap.
But when I want to change bitmap to ALPHA_8, the [Cue encode is error](https://github.com/androidx/media/issues/2054), and `SubtitlePainter` also need [Blend bitmap](https://github.com/androidx/media/issues/2055).

Why do I stick with this approach?

Because this is a most simple way to render ass with libass, I don't need to worry about subtitles syncing with the video，and compat with current sutitle render process，with a few change in cue and SubtitlePainter.

If we use Effect (If it can work), there is more work I need todo, and to hack exo extractor and prase process.

**Comment by moi15moi** [comment]
@icbaker I'm currently trying to add `LibassSubtitleRenderer`, which would use libass, and eventually open a PR.  
As you already mentioned, the `Cue` machinery doesn't "fit" how libass works.  

If I am not mistaken, ASS/SSA subtitles are only available for the Matroska container, so all the decisions I made were based on this observation.  

| Libass Method | Media3 Function |
|--------------|----------------|
| [`ass_render_frame`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L609-L618) | [`ExoPlayer.setVideoEffects()`](https://developer.android.com/reference/androidx/media3/exoplayer/ExoPlayer#setVideoEffects(java.util.List%3Candroidx.media3.common.Effect%3E)) + [`BitmapOverlay`](https://developer.android.com/reference/androidx/media3/effect/BitmapOverlay). I would return the bitmap generated by libass in [`BitmapOverlay.getBitmap()`](https://developer.android.com/reference/androidx/media3/effect/BitmapOverlay#getBitmap(long)). Also, in the bitmap alpha blending, I would call [`ExoPlayer.getVideoFormat().colorInfo`](https://developer.android.com/reference/androidx/media3/exoplayer/ExoPlayer#getVideoFormat()) to do the "VSFilter mangled colors" explained [here](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass_types.h#L157-L262)  |
| [`ass_set_frame_size`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L385-L401) | Use the value obtained from [`Player.getSurfaceSize()`](https://developer.android.com/reference/androidx/media3/common/Player#getSurfaceSize()) + [`Player.Listener.onSurfaceSizeChanged()`](https://developer.android.com/reference/androidx/media3/common/Player.Listener#onSurfaceSizeChanged(int,int)). |
| [`ass_set_storage_size`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L403-L421) | Use the value obtained from [`Player.getVideoSize()`](https://developer.android.com/reference/androidx/media3/common/Player#getVideoSize()) + [`Player.Listener.onVideoSizeChanged()`](https://developer.android.com/reference/androidx/media3/common/Player.Listener#onVideoSizeChanged(androidx.media3.common.VideoSize)). |
| [`ass_process_codec_private`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L707-L714) | [`SsaParser` constructor](https://github.com/androidx/media/blob/44b3a43652c0fb462cd1799d253d663e5207d268/libraries/extractor/src/main/java/androidx/media3/extractor/text/ssa/SsaParser.java#L108), but `SsaParser` implements `SubtitleParser`, which requires me to use `Cues`, which I cannot do. So I'm not sure what I should do. |
| [`ass_process_chunk`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L716-L731) | [`TrackOutput.sampleMetadata()`](https://developer.android.com/reference/androidx/media3/extractor/TrackOutput#sampleMetadata(long,int,int,int,androidx.media3.extractor.TrackOutput.CryptoData)) seems to receive the Matroska block, but [`media3`](https://github.com/androidx/media/blob/44b3a43652c0fb462cd1799d253d663e5207d268/libraries/extractor/src/main/java/androidx/media3/extractor/mkv/MatroskaExtractor.java#L1491) seems to modify the block data and add the duration. It should be fine, but it is something I need to verify. |
| [`ass_add_font`](https://github.com/libass/libass/blob/1b699559025185e34d21a24cac477ca360cb917d/libass/ass.h#L806-L814) *(Libass absolutely needs the font muxed into the MKV)* | It is fairly simple to extract the font from the MKV (I did it [here](https://github.com/moi15moi/media/pull/1/files)), but I don't see how I can pass the fonts to libass. |

I saw your comment [here](https://github.com/androidx/media/issues/2053#issuecomment-2601855558):

> This would involve a custom `ExtractorsFactory` and a customized `ForwardingExtractorOutput` that lets you intercept the subtitle data.

By default, Media3 uses [`DefaultExtractorsFactory`](https://github.com/androidx/media/blob/44b3a43652c0fb462cd1799d253d663e5207d268/libraries/extractor/src/main/java/androidx/media3/extractor/DefaultExtractorsFactory.java#L489), so I assume I would need to modify this class.  

How could I add a special case to use libass in this class?  

Also, I don't quite understand what I should edit, so [this line](https://github.com/androidx/media/blob/44b3a43652c0fb462cd1799d253d663e5207d268/libraries/extractor/src/main/java/androidx/media3/extractor/mkv/MatroskaExtractor.java#L2456) returns my custom `TrackOutput`. Could you give me some detail?

**Comment by icbaker** [comment]
> and eventually open a PR.

Just to set expectations, based on the sketched design in your comment, I'm afraid we are unlikely to merge such a PR.

The design relies on media3's `lib-effect`, and `libass` native code, neither of which are currently depended on by `lib-exoplayer`, so either we'd need to add these deps (which then affects **every** user of the player library), or this would need to be in a new `lib-ssa-parser` module or similar, which comes with significant maintenance burden that I'm afraid probably isn't justified.

The design being so 'different' to all other subtitle parsers (by necessity) also increases the maintenance burden.

So I'm afraid a design like this will have to exist as an extension maintained outside the main media3 repository.

**Comment by peerless2012** [comment]
Support extract fonts in mkv and pass them to SubtitleParser maybe available?

In our project, we use [AssMatroskaExtractor](https://github.com/peerless2012/libass-android/blob/master/lib_ass_media/src/main/java/io/github/peerless2012/ass/media/extractor/AssMatroskaExtractor.kt) 
witch  extend MatroskaExtractor to extract the fonts attachments.

**Comment by e1ker** [comment]
> So I'm afraid a design like this will have to exist as an extension maintained outside the main media3 repository.

I'm not sure if this means it wouldn't be officially integrated into the player, but if that's the case even an unofficial fork would be great. I feel this is the only thing missing to have a player that "just works" not only on Android but on almost every other OS, since most have very big compromises, are too heavy or too hard to set up.

**Comment by peerless2012** [comment]
> > So I'm afraid a design like this will have to exist as an extension maintained outside the main media3 repository.
> 
> I'm not sure if this means it wouldn't be officially integrated into the player, but if that's the case even an unofficial fork would be great. I feel this is the only thing missing to have a player that "just works" not only on Android but on almost every other OS, since most have very big compromises, are too heavy or too hard to set up.

There is already a third plugin to bring libass in media3, you just need to add only one method to your players construct.

Here's an example:
```
player = ExoPlayer.Builder(this)
            .buildWithAssSupport(
                this,
                AssRenderType.OPEN_GL
            )
```

See [buildWithAssSupport](https://github.com/peerless2012/libass-android/blob/2a3dbb48c6c05de0ba14027dd13f7ba58e572b7a/app/src/main/java/io/github/peerless2012/ass/demo/MainActivity.kt#L44)

**Comment by GlassedSilver** [comment]
> > and eventually open a PR.
> 
> Just to set expectations, based on the sketched design in your comment, I'm afraid we are unlikely to merge such a PR.
> 
> The design relies on media3's `lib-effect`, and `libass` native code, neither of which are currently depended on by `lib-exoplayer`, so either we'd need to add these deps (which then affects **every** user of the player library), or this would need to be in a new `lib-ssa-parser` module or similar, which comes with significant maintenance burden that I'm afraid probably isn't justified.
> 
> The design being so 'different' to all other subtitle parsers (by necessity) also increases the maintenance burden.
> 
> So I'm afraid a design like this will have to exist as an extension maintained outside the main media3 repository.

Honest question from someone very grateful exoplayer exists to begin with <3

Why is it unreasonable to include those dependencies?

If the goal is to support more subtitles which is a goal of accessibility I don't see how that's something that shouldn't be universal.


And before anyone says it... Text-only subs that don't rely on styling are not always an equal but not as pretty path. Quite frequently subs need to be placed in various different places and styling is often done to differentiate between characters, underline the mood/atmosphere /etc...

**Comment by nigelmegitt** [comment]
> Text-only subs that don't rely on styling are not always an equal but not as pretty path. Quite frequently subs need to be placed in various different places and styling is often done to differentiate between characters, underline the mood/atmosphere /etc...

ExoPlayer already supports other formats than SSA/ASS that allow for positioning and styling. The argument for SSA/ASS is, as I understand it, to do with specific styling options that are only available in that format and that are particularly popular with some genres of content.

**Comment by GlassedSilver** [comment]
> > Text-only subs that don't rely on styling are not always an equal but not as pretty path. Quite frequently subs need to be placed in various different places and styling is often done to differentiate between characters, underline the mood/atmosphere /etc...
> 
> ExoPlayer already supports other formats than SSA/ASS that allow for positioning and styling. The argument for SSA/ASS is, as I understand it, to do with specific styling options that are only available in that format and that are particularly popular with some genres of content.

Correct, however when all that you have is SSA/ASS for a given piece of content (at least properly detailed subs, a lot of times rather mediocre compromise subs exist in text-only) then wishing for the other options getting used will get you nowhere.

How many genres are affected doesn't change that this does make some content less accessible.

**Comment by sam-reza** [comment]
...

**Comment by kayodde** [comment]
I hope there are ways to get exoplayer to have proper as subtitles soon especially on Android tv. I use Artemio and exoplayer can play far more videos than libvlc, however the ass subtitles are poor and I can't read them as they are white with no borders. And their "ass support" feature does not work. Are there any exo based external players that have these features?
